File: packages/discovery-provider/src/utils/rendezvous.py
from hashlib import sha256
from typing import List

dead_nodes = ["https://content.grassfed.network/"]


# Python equivalent of https://github.com/tysonmote/rendezvous/blob/be0258dbbd3d/rendezvous.go
class RendezvousHash:
    def __init__(self, *nodes: str):
        self.nodes: List[str] = [n for n in nodes if n not in dead_nodes]

    def add(self, *nodes: str) -> None:
        for node in nodes:
            if node not in dead_nodes:
                self.nodes.append(node)

    def get_nodes(self) -> List[str]:
        return self.nodes

    def get(self, key: str) -> str:
        ranked = self.rank_sha256(key)
        if ranked:
            return ranked[0]
        return ""

    def get_n(self, n: int, key: str) -> List[str]:
        return self.rank_sha256(key)[:n]

    def rank_sha256(self, key: str) -> List[str]:
        tuples = [
            (h, sha256((h + key).encode("utf-8")).hexdigest()) for h in self.nodes
        ]
        tuples.sort(key=lambda t: (t[1], t[0]))
        return [t[0] for t in tuples]

File: packages/sdk/src/sdk/utils/rendezvous.ts
import { Buffer } from 'buffer'

import { sha256 } from '@noble/hashes/sha256'
import { bytesToHex as toHex } from '@noble/hashes/utils'

class NodeScore {
  node: Buffer
  score: number

  constructor(node: Buffer) {
    this.node = node
    this.score = 0
  }
}

/**
 * TypeScript equivalent of https://github.com/tysonmote/rendezvous/blob/be0258dbbd3d/rendezvous.go
 */
export class RendezvousHash {
  private readonly nodes: NodeScore[] = []

  constructor(...nodes: string[]) {
    this.add(...nodes)
  }

  add(...nodes: string[]): void {
    for (const node of nodes) {
      this.nodes.push(new NodeScore(Buffer.from(node)))
    }
  }

  getNodes(): string[] {
    return this.nodes.map((nodeScore) => nodeScore.node.toString())
  }

  get(key: string): string {
    const first = this.getN(1, key)[0]
    return first ?? ''
  }

  getN(n: number, key: string): string[] {
    return this.rendezvous256(key).slice(0, n)
  }

  rendezvous256(key: string) {
    const tuples = this.nodes.map((n) => {
      const hostName = n.node.toString()
      return [hostName, toHex(sha256(`${hostName}${key}`))]
    })
    tuples.sort((t1, t2) => {
      const [aHost, aScore] = t1
      const [bHost, bScore] = t2
      if (aScore === bScore) {
        return aHost! < bHost! ? -1 : 1
      }
      return aScore! < bScore! ? -1 : 1
    })
    return tuples.map((t) => t[0]!)
  }
}
